src/Database/LSMTree/Internal/Serialise/Class.hs

```diff
@@ -176,59 +182,71 @@ instance SerialiseValue String where
 
--- | Placeholder instance, not optimised
+-- | \( O(n) \) (de-)serialisation, where \(n\) is the number of bytes
+--
+-- TODO: optimise, it's \( O(n) + O(n) \) where it could be \( O(n) \).
 instance SerialiseKey LBS.ByteString where
   serialiseKey = serialiseKey . LBS.toStrict
   deserialiseKey = B.toLazyByteString . RB.builder
 
--- | Placeholder instance, not optimised
-instance SerialiseKey BS.ByteString where
-  serialiseKey = RB.fromShortByteString . SBS.toShort
-  deserialiseKey = LBS.toStrict . deserialiseKey
+instance SerialiseKeyOrderPreserving LBS.ByteString
 
--- | Placeholder instance, not optimised
+-- | \( O(n) \) (de-)serialisation, where \(n\) is the number of bytes
+--
+-- TODO: optimise, it's \( O(n) + O(n) \) where it could be \( O(n) \).
 instance SerialiseValue LBS.ByteString where
   serialiseValue = serialiseValue . LBS.toStrict
   deserialiseValue = B.toLazyByteString . RB.builder
 
--- | Placeholder instance, not optimised
-instance SerialiseValue BS.ByteString where
-  serialiseValue = RB.fromShortByteString . SBS.toShort
-  deserialiseValue = LBS.toStrict . deserialiseValue
+-- | \( O(n) \) (de-)serialisation, where \(n\) is the number of bytes
+instance SerialiseKey BS.ByteString where
+  serialiseKey = serialiseKey . SBS.toShort
+  deserialiseKey = SBS.fromShort . deserialiseKey
 
-{-------------------------------------------------------------------------------
- ShortByteString
--------------------------------------------------------------------------------}
+instance SerialiseKeyOrderPreserving BS.ByteString
+
+-- | \( O(n) \) (de-)serialisation, where \(n\) is the number of bytes
+instance SerialiseValue BS.ByteString where
+  serialiseValue = serialiseValue . SBS.toShort
+  deserialiseValue = SBS.fromShort . deserialiseValue
 
+-- | \( O(1) \) (de-)serialisation
 instance SerialiseKey SBS.ShortByteString where
   serialiseKey = RB.fromShortByteString
   deserialiseKey = byteArrayToSBS . RB.force
 
+instance SerialiseKeyOrderPreserving SBS.ShortByteString
+
+-- | \( O(1) \) (de-)serialisation
 instance SerialiseValue SBS.ShortByteString where
   serialiseValue = RB.fromShortByteString
   deserialiseValue = byteArrayToSBS . RB.force
 
 {-------------------------------------------------------------------------------
- ByteArray
+  ByteArray
 -------------------------------------------------------------------------------}
 
--- | The 'Ord' instance of 'ByteArray' is not lexicographic, so there cannot be
--- an order-preserving instance of 'SerialiseKey'.
--- Use 'ShortByteString' instead.
+-- | \( O(1) \) (de-)serialisation
+instance SerialiseKey P.ByteArray where
+  serialiseKey ba = RB.fromByteArray 0 (P.sizeofByteArray ba) ba
+  deserialiseKey = RB.force
+
+-- | \( O(1) \) (de-)serialisation
 instance SerialiseValue P.ByteArray where
   serialiseValue ba = RB.fromByteArray 0 (P.sizeofByteArray ba) ba
   deserialiseValue = RB.force
```

Comment:
Oops, `RB.force` is $O(n)$ instead of $O(1)$. Serialisation is still $O(1)$, but deserialisation is not.